ffi/derive/src/lib.rs

```diff
@@ -1,47 +1,176 @@
+        let items = &self.0;
+        tokens.extend(quote! {#(#items)*})
+    }
+}
+
+/// Derive implementations of traits required to convert to and from an FFI-compatible type
+#[proc_macro]
+#[proc_macro_error::proc_macro_error]
+pub fn ffi(input: TokenStream) -> TokenStream {
+    let items = parse_macro_input!(input as FfiItems).0;
+
+    #[cfg(feature = "client")]
+    let items = items.iter().map(|item| {
+        if is_opaque(item) {
+            wrapper::wrap_as_opaque(item)
+        } else {
+            quote! {#item}
+        }
+    });
+
+    quote! { #(#items)* }.into()
+}
+
+/// Derive implementations of traits required to convert to and from an FFI-compatible type
+#[proc_macro_derive(IntoFfi)]
+#[proc_macro_error::proc_macro_error]
+pub fn into_ffi_derive(input: TokenStream) -> TokenStream {
+    let item = parse_macro_input!(input as syn::DeriveInput);
+    let into_ffi_derive = derive_into_ffi(&item);
+
+    if !matches!(item.vis, syn::Visibility::Public(_)) {
+        abort!(item.vis, "Only public items are supported");
+    }
+
+    if !item.generics.params.is_empty() {
+        abort!(item.generics, "Generics are not supported");
+    }
+
+    quote! { #into_ffi_derive }.into()
+}
+
+/// Derive implementation of [`TryFromReprC`] trait
+#[proc_macro_derive(TryFromReprC)]
+#[proc_macro_error::proc_macro_error]
+pub fn try_from_repr_c_derive(input: TokenStream) -> TokenStream {
+    let item = parse_macro_input!(input as syn::DeriveInput);
+    let try_from_ffi_derive = derive_try_from_ffi(&item);
+
+    if !matches!(item.vis, syn::Visibility::Public(_)) {
+        abort!(item.vis, "Only public items are supported");
+    }
+
+    if !item.generics.params.is_empty() {
+        abort!(item.generics, "Generics are not supported");
+    }
+
+    quote! { #try_from_ffi_derive }.into()
+}
 
 /// Generate FFI functions
+///
+/// # Example:
+/// ```no_run
+/// use getset::Getters;
+/// use iroha_ffi::{slice::OutSliceRef, FfiReturn, IntoFfi};
+///
+/// // For a struct such as:
+/// #[derive(Getters, IntoFfi)]
+/// #[getset(get = "pub")]
+/// pub struct Foo {
+///     /// Id of the struct
+///     id: u8,
+///     #[getset(skip)]
+///     bar: Vec<u8>,
+/// }
+///
+/// impl Foo {
+///     /// Construct new type
+///     fn new(id: u8) -> Self {
+///         Self {id, bar: Vec::new()}
+///     }
+///     /// Return bar
+///     fn bar(&self) -> &[u8] {
+///         &self.bar
+///     }
+/// }
+///
+/// // The following functions will be derived:
+/// extern "C" fn Foo__new(id: u8, output: *mut Foo) -> FfiReturn {
+///     /* function implementation */
+///     FfiReturn::Ok
+/// }
+/// extern "C" fn Foo__bar(handle: *const Foo, output: OutSliceRef<u8>) -> FfiReturn {
+///     /* function implementation */
+///     FfiReturn::Ok
+/// }
+/// extern "C" fn Foo__id(handle: *const Foo, output: *mut u8) -> FfiReturn {
+///     /* function implementation */
+///     FfiReturn::Ok
+/// }
+/// ```
```

Comment:
One moment. How does you macro can generate `Foo__bar` if bar was skipped with `getset(skip)` and was implemented inside `impl Foo`? You macro simply cannot obtain information about `impl Foo` block...

Reply:
should be fixed, it was the intention to derive from the `impl` block